File: ABM-SIM-DOTNET/src/Moment.cs
using MathNet;
using MathNet.Numerics.Statistics;

namespace ABMSIM {
    public class Moment {
        public readonly double mean;
        public readonly double stddev;
        public readonly float midPrice;
        public readonly float spread;

        public readonly float p_k;
        public readonly float bestOffer;
        public readonly float bestBid;

        public Moment(OrderBook ob) {
            List<float> Bidprices = ob.bids.Select(o => o.price).ToList();
            List<float> Askprices = ob.asks.Select(o => o.price).ToList();

            mean = Bidprices.Average();
            stddev = Bidprices.StandardDeviation();
            midPrice = Bidprices.Median();

            bestBid = Askprices.Min();
            bestOffer = Bidprices.Max();
            spread = Math.Abs(bestBid - bestOffer);
        }
    }
}

File: abm-sim-js/src/sim/Moment.js
import {std, mean, median, max, min} from "mathjs"

export default class Moment {
  mean 
  stddev 
  midPrice
  spread
  p_k 
  min
  max
  constructor(ob) {
    this.mean = mean(ob.prices)
    this.stddev = std(ob.prices)
    this.midPrice = median(ob.prices)
    
    this.max = max(ob.prices)
    this.min = min(ob.prices)
    this.spread = Math.abs(this.max - this.min)

    this.p_k = ob.asks.length + ob.bids.length > 0 ? (ob.asks.length - ob.bids.length) / ob.asks.length + ob.bids.length : 0 
  }

  ToString() {
    return `{
mean: ${this.mean}
midPrice: ${this.midPrice}
spread: ${this.spread}
p_k: ${this.p_k}
min: ${this.min}
max: ${this.max} 
}`
  }
}